### DISCUSSION TITLE
DIDComm / Veramo over libp2p?

### DISCUSSION THREAD

**Discussion by bmiller59** [decentralized-identity/veramo#769] [Q&A]
Hello. Recently read your exciting article about [DIDComm over libp2p](https://medium.com/uport/didcomm-messaging-through-libp2p-cffe0f06a062). 

Can you provide a simple MVP demonstration of using Veramo over libp2p to 
* Pass public, signed DIDComm broadcast messages across a libp2p network
* Pass private (encrypted), signed DIDComm messages across a libp2p network

If you can also speak to your medium/long-term plans for releasing open source extending the uses of Veramo over libp2p that would be very helpful. 

Thank you for your good work!

PS - You say to watch your Veramo blog for updates, but it offers no RSS feed and no way to subscribe over email. You may want to consider offering one or the other. ;) Thx.

**Comment by alenhorvat** [comment]
Hi. 

We are actively working on a demo and we'll notify you as soon as we complete it.

**Comment by alenhorvat** [comment]
Hi, 
an early version of the demo is available 
(https://bitbucket.org/netis/veramo-aceblock-didcomm-over-libp2p-react/src/master/)

Several improvements will follow (on the documentation and code level).

**Reply by bmiller59** [reply]
@alenhorvat Thank you for the demo! I am trying to use it by setting up two different users on the same machine. But I keep getting the error `failed to dial to remote` with reason `dial to self attempted`. Can you help me resolve this so I can test out the demo?

* First I tried using two different accounts in two different browsers on the same wallet and received this error.
* Then I tried using two totally separate wallets in two different browser and got the same error.
* Finally I cloned the project repo twice and initialized them to two different port sets as described in the readme, then opened them with two separate wallets in two browsers and still got the same error.

I am out of other things to try. Can you provide some guidance how to resolve the issue and explain what "self" refers to in this case? Much appreciated!

**Reply by alenhorvat** [reply]
Hi. 

As this is a very simple demo, you need to have two copies (as you mention in the 3rd point)
Please check that
- in backend/config/client.json private keys of two instances are different
- you changed ports in the following files: frontend/.env, backend/.env, "proxy" variable in frontend/package.json

`dial to self attempted` is a libp2p warning and it occurs when a libp2p client tries to dial itself.

Another potential issue is that the two accounts you're using have the same libp2p service endpoint.

If the issue persists, let me know.

**Reply by bmiller59** [reply]
@alenhorvat Thanks! I had forgotten to change the proxy in the package.json file. That was the problem. Now I am able to send messages.

**Reply by alenhorvat** [reply]
The demo uses simple JWS messages, however, any packaging supported by the Veramo can be used.
About _Pass public, signed DIDComm broadcast messages across a libp2p network_
do you mean by using a 1-to-many (fan-out) approach or using something pub-sub? (both approaches are possible)

**Reply by elribonazo** [reply]
I'm curious because idea of didcomm is rotating the dids as much as possible, how have u guys solved that in the libp2p node? 
I got really inspired by this @alenhorvat :) Thank you!! 

Proly I'll find out soon but i'm having some trouble creating a peerID from a set of valid ed25519Keys. The PeerID creates and node starts but the Noise encryption breaks in Handshake. Any idea?

**Reply by alenhorvat** [reply]
Hi @elribonazo. Note that Node ID and didcomm keys are used for a different purpose. 
Node ID doesn't have to be rotated often since message encryption (also) happens on the didcomm level.

Depending on your use case, you can come up with a node key rotation strategy. There are several options, one being signal protocol-like double ratchet or simply pre-sharing authenticated public keys. Happy to help with the topic, if needed.

(note that we started working on a PoC where signal protocol is used over libp2p)

Issue you're having may be related to the libp2p lib itself.

**Reply by elribonazo** [reply]
Hello @alenhorvat,
WOW, thanks for the quick answer !

_Node ID doesn't have to be rotated often since message encryption (also) happens on the DIDComm level._
I get that, but then why sending a DIDComm message to a Peer if you really don't know if he will be able to decrypt it, I get what u mean though.

_Issue you're having may be related to the libp2p lib itself._
"Sadly yes", I'm doing a workaround atm though and just generating random peerId's before the node starts, but my main goal is to be able to instantiate the PeerID that I'll use to start the node using a pre existing ed25519 keyPair. Maybe u guys have gone through this, hoping to :) 

On any case again, thank you soo much for the quick answer and the hard work u guys do!
